### PR TITLE
Set stackTraceLimit to 0 in fileSystemEntryExists 

### PR DESCRIPTION
**Pull request by gluxon**
The exception thrown by Node.js's fs.statSync function contains a stack trace that can be expensive to compute. Since this exception isn't used by fileSystemEntryExists, we can safely set Error.stackTraceLimit to 0 without a change in behavior.

A significant performance improvement was noticed with this change while profiling tsserver on packages within a proprietary monorepo. Specifically, my team saw high self time percentages for Node.js's `uvException` and `handleErrorFromBinding` internal functions. These functions are executed within `fs.statSync` when it fails to find the given path.

Screen Shot 2020-08-12 at 4 13 41 PM

`fs.statSync`: https://github.com/nodejs/node/blob/v14.4.0/lib/fs.js#L1030-L1037
`handleErrorFromBinding`: https://github.com/nodejs/node/blob/v14.4.0/lib/internal/fs/utils.js#L254-L269
`uvException`: https://github.com/nodejs/node/blob/v14.4.0/lib/internal/errors.js#L390-L443

 ## Measurements

After adding `Error.stackTraceLimit = 0`, we saw:

- For a large configured project with 12,565 files, tsserver reached the `projectLoadingFinish` event 48.78% faster. (~46.786s vs ~31.447s)
- For a medium project with 7,064 files, tsserver was 25.75% faster. (~20.897s vs ~16.618s)
- For a small project with 796 files, tsserver was only a negligible 3.00% faster. (~3.545s vs ~3.442)

Measurements were taken on macOS 10.15.6, Node.js 14.4.0, and a recent master commit of TypeScript (`610fa28d`). The average of 3 runs before and after this change were taken.

I would normally include `.cpuprofile` and `isolate-*-*-*.log` files, but can't post them publicly in this case. If there's any other summaries the TypeScript team would be curious about I can report them.

## fs.statSync Misses

Within our monorepo, the `fs.statSync` misses were mostly searches for alternative file extensions of module imports.

- For node_modules imports, a lot of `.ts`/`.tsx` lookups failed until the `.d.ts` file was found.
- Within projects with a lot of JSX files, `.ts` files were looked for before finding the `.tsx` version.
- In the medium scale project mentioned above, a total of 38,515 non-existent files were queried during `createProgram`.

### CODE REVIEW

**Comment by gluxon** [comment]
Tagging @amcasey due to the relation with PR #36190.

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [X] Code is up-to-date with the `master` branch
* [X] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

New tests may be a bit difficult with this specific change. Open to any testing suggestions if anyone has them. Also not sure if what the associated backlog issue for this should be.

File: src/compiler/sys.ts [hunk 1]
```diff
@@ -1662,6 +1662,12 @@ namespace ts {
             }
 
             function fileSystemEntryExists(path: string, entryKind: FileSystemEntryKind): boolean {
+                // fs.statSync can be slow if path doesn't exist since Node.js generates a stack trace. Adding an
+                // fs.existsSync guard here can improve performance by reducing the number of times fs.statSync throws.
+                if (!_fs.existsSync(path)) {
+                    return false;
+                }
+
                 try {
```
Comment (sheetalkamat) [inline src/compiler/sys.ts:1671]: Another option is to just return using existsSync and ignore directory or file and make our code more tolerant (it might already be) for finding file instead of directory or vice versa. That could give us better perf in both cases.

Comment (gluxon) [reply src/compiler/sys.ts:1671]: Definitely. I was also wondering how necessary it was to check for file versus folder existence.

I'm guessing the current behavior of this function is this way so folders ending with `.ts` get ignored. Is that an edge case where a behavior change would be acceptable?

**Comment by sheetalkamat** [comment]
Not sure but if you make a change, we can run user tests and see what breaks

File: src/compiler/sys.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/sys.ts:1671]: If you make a change we can see what user tests break and see what happens

**Comment by gluxon** [comment]
I ended up removing the `!fs.existsSync` check favoring `Error.stackTraceLimit = 0` instead. This is a pattern Node.js core uses as well: https://github.com/nodejs/node/blob/v14.4.0/lib/internal/errors.js#L414-L423 I reran some timings and this solution is comparable to the original in terms of performance gains.

Since the happy path (when the path exists) no longer calls both `fs.existsSync` and `fs.statSync`, I also removed the "Other Notes" caveat from the PR description.

**Comment by amcasey** [comment]
@gluxon, it's precisely the file/dir distinction that prevented me from rolling out `existsSync` more broadly.  It's possible we can add try-catches for the rare cases where the wrong type is found, but it seemed likely to be a pervasive change.

As an alternative, I've proposed a change to node itself: https://github.com/nodejs/node/pull/33716

**Comment by amcasey** [comment]
Without having re-reviewed the code, I'm also somewhat concerned that the exists/non-exists balance might be different in long-running watch or language service scenarios than it is during batch compilation.

File: src/compiler/sys.ts (hunk 1 above)
Comment (gluxon) [reply src/compiler/sys.ts:1670]: Ended up not doing this and favoring `Error.stackTraceLimit = 0` instead. I assume we wanted to try using `existsSync` entirely since we were worried about perf in the path exists case. Setting `Error.stackTraceLimit = 0` should improve performance when files don't exist with no change to checking files that do exist.

**Comment by gluxon** [comment]
> As an alternative, I've proposed a change to node itself: nodejs/node#33716

That's a great proposed change. I was quite frustrated that there's no fast way to `stat` in node. Thanks for opening that.

> Without having re-reviewed the code, I'm also somewhat concerned that the exists/non-exists balance might be different in long-running watch or language service scenarios than it is during batch compilation.

I ended up using the `Error.stackTraceLimit = 0` trick mentioned in your link and isolated it to this specific hotspot. This change should be insensitive to the exists/non-exists balance.

**Comment by gluxon** [comment]
It looks like the node12 and node14 tests timed out cloning this PR. Pushing an amended commit to restart CI.

File: src/compiler/sys.ts (hunk 1 above)
Comment (amcasey) [reply src/compiler/sys.ts:1670]: Nice!

**Comment by gluxon** [comment]
> Love the tailored mitigation!

Awesome. Thanks to you and @sheetalkamat for the fast reviews!

**Comment by sheetalkamat** [comment]
@typescript-bot perf test this

**Comment by ericjeney** [comment]
Would it be possible to backport this to the 4.0 branch as well, once merged? My team is eager to get this performance improvement as soon as possible.

**Comment by amcasey** [comment]
@ericjeney What kind of improvement are you seeing?  When I was investigating (before this PR), it looked like it was in the 5% range, which is certainly very nice, but is not hold-the-release.  Having data from real customers would make it easier to argue for a late change.

**Comment by ericjeney** [comment]
@gluxon and I are on the same team, so my results are similar to his. We work on a code base of ~1,500,000 lines of TypeScript spread across ~500 packages.

In addition to the benchmarks he noted above, I tested on a patched version of `tsserver.js` in VS Code and timed the "Initializing JS/TS Language Features..." prompt after selecting the Restart TS Server action. I noted a ~20% improvement for several packages, going from a consistent 38s => 28-30s for one of them.

If not in 4.0, is this the sort of change you would ship in a patch release? Or are patch releases reserved for regressions?

**Comment by amcasey** [comment]
@ericjeney We'll have to discuss the potential risks, but those numbers are pretty exciting.  I'll post more when I know more.

**Comment by ericjeney** [comment]
Thank you!

I just confirmed with another coworker as well, who works on a separate monorepo, and saw a >25% improvement on one of their packages (17.5s => 13s), so the improvement seems to not be isolated to only the codebase that I work on.

**Comment by amcasey** [comment]
@ericjeney If things look good in nightly builds, this will be a strong candidate for inclusion in a patch release.  No promises though.

**Comment by gluxon** [comment]
Thanks for the update @amcasey! Is there anything I still need to do on our side before this can merge?

**Comment by amcasey** [comment]
@gluxon, nope, just wanted to hear from @sheetalkamat. :smile:  Thanks again!

**Comment by ericanderson** [comment]
Thanks @amcasey / @weswigham / @sheetalkamat !

**Comment by amcasey** [comment]
@typescript-bot cherry-pick this to release-4.0
